src/backend/parser/parse_coerce.c

```diff
@@ -1437,8 +1439,9 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
 				pcategory = ncategory;
 				pispreferred = nispreferred;
 			}
-			else if (ncategory != pcategory
-				&& sql_dialect != SQL_DIALECT_TSQL) /* T-SQL allows to select common datatype between different categories */
+			else if (ncategory != pcategory &&
+				sql_dialect != SQL_DIALECT_TSQL && /* T-SQL allows to select common datatype between different categories */
+				(!dump_restore || (dump_restore && strcmp(dump_restore, "on") != 0))) /* allow common datatype between different categories while restoring babelfish database */
```

Comment:
Besides this case, am just wondering if there are any test cases for the additions below.

Reply:
Actually all the changes as a whole fixes the issue completely. The above condition will just skip the error when there are datatypes from different categories but in the `if` condition just below this, we will call `determine_datatype_precedence_hook` and `can_coerce_type` functions, `can_coerce_type` will eventually call `find_coercion_pathway_hook` hook to search into T-SQL coercion rules. But I will anyway add few more similar test cases to ensure we don't miss any edge case.